File: backend/app/database/connection.py
import sqlite3
from pathlib import Path
from contextlib import contextmanager

DB_PATH = Path(__file__).resolve().parent.parent.parent / "assetops.db"

def get_db_connection(db_path: Path = DB_PATH):
    conn = sqlite3.connect(str(db_path), check_same_thread=False)
    conn.row_factory = sqlite3.Row
    conn.execute("PRAGMA foreign_keys = ON")
    conn.execute("PRAGMA journal_mode = WAL")
    return conn

def get_db(db_path: Path = DB_PATH):
    conn = get_db_connection(db_path)
    try:
        yield conn
        conn.commit()
    except Exception:
        conn.rollback()
        raise
    finally:
        conn.close()

def init_db(db_path: Path = DB_PATH):
    conn = get_db_connection(db_path)
    with conn:
        conn.executescript("""
        CREATE TABLE IF NOT EXISTS assets (
            address TEXT PRIMARY KEY,
            name TEXT NOT NULL,
            symbol TEXT NOT NULL,
            decimals INTEGER NOT NULL DEFAULT 18,
            total_supply TEXT NOT NULL DEFAULT '0',
            status TEXT NOT NULL DEFAULT 'ACTIVE',
            created_at TEXT NOT NULL DEFAULT (datetime('now'))
        );

        CREATE TABLE IF NOT EXISTS holders (
            id TEXT PRIMARY KEY,
            address TEXT NOT NULL,
            asset_address TEXT NOT NULL,
            balance TEXT NOT NULL DEFAULT '0',
            is_whitelisted INTEGER NOT NULL DEFAULT 1,
            updated_at TEXT NOT NULL DEFAULT (datetime('now')),
            UNIQUE(address, asset_address),
            FOREIGN KEY (asset_address) REFERENCES assets(address)
        );

        CREATE TABLE IF NOT EXISTS transfers (
            id TEXT PRIMARY KEY,
            asset_address TEXT NOT NULL,
            from_address TEXT NOT NULL,
            to_address TEXT NOT NULL,
            amount TEXT NOT NULL,
            timestamp TEXT NOT NULL DEFAULT (datetime('now')),
            FOREIGN KEY (asset_address) REFERENCES assets(address)
        );

        CREATE TABLE IF NOT EXISTS corporate_actions (
            action_id TEXT PRIMARY KEY,
            action_reference TEXT NOT NULL,
            asset_address TEXT NOT NULL,
            action_type TEXT NOT NULL,
            active_version_id TEXT NOT NULL,
            status TEXT NOT NULL DEFAULT 'ACTIVE',
            created_at TEXT NOT NULL DEFAULT (datetime('now')),
            FOREIGN KEY (asset_address) REFERENCES assets(address)
        );

        CREATE TABLE IF NOT EXISTS action_versions (
            version_id TEXT PRIMARY KEY,
            action_id TEXT NOT NULL,
            version_number INTEGER NOT NULL,
            rate_bps INTEGER,
            amount_per_token TEXT,
            record_date TEXT,
            payable_date TEXT,
            supersedes_version_id TEXT,
            document_hash TEXT,
            status TEXT NOT NULL DEFAULT 'ACTIVE',
            created_at TEXT NOT NULL DEFAULT (datetime('now')),
            FOREIGN KEY (action_id) REFERENCES corporate_actions(action_id),
            UNIQUE(action_id, version_number)
        );

        CREATE TABLE IF NOT EXISTS payment_executions (
            id TEXT PRIMARY KEY,
            action_id TEXT NOT NULL,
            version_id TEXT NOT NULL,
            total_paid TEXT NOT NULL,
            executed_at TEXT NOT NULL DEFAULT (datetime('now')),
            FOREIGN KEY (action_id) REFERENCES corporate_actions(action_id),
            FOREIGN KEY (version_id) REFERENCES action_versions(version_id),
            UNIQUE(version_id)
        );

        CREATE TABLE IF NOT EXISTS payment_events (
            id TEXT PRIMARY KEY,
            action_id TEXT NOT NULL,
            version_id TEXT NOT NULL,
            holder_address TEXT NOT NULL,
            amount TEXT NOT NULL,
            asset_balance TEXT NOT NULL,
            timestamp TEXT NOT NULL DEFAULT (datetime('now')),
            FOREIGN KEY (action_id) REFERENCES corporate_actions(action_id),
            FOREIGN KEY (version_id) REFERENCES action_versions(version_id)
        );

        CREATE TABLE IF NOT EXISTS redemption_events (
            id TEXT PRIMARY KEY,
            action_id TEXT NOT NULL,
            version_id TEXT NOT NULL,
            holder_address TEXT NOT NULL,
            tokens_burned TEXT NOT NULL,
            principal_amount TEXT NOT NULL,
            timestamp TEXT NOT NULL DEFAULT (datetime('now')),
            FOREIGN KEY (action_id) REFERENCES corporate_actions(action_id),
            FOREIGN KEY (version_id) REFERENCES action_versions(version_id)
        );

        CREATE TABLE IF NOT EXISTS audit_logs (
            event_id TEXT PRIMARY KEY,
            event_type TEXT NOT NULL,
            action_id TEXT,
            version_id TEXT,
            actor TEXT,
            details TEXT NOT NULL,
            prev_hash TEXT,
            event_hash TEXT NOT NULL,
            timestamp TEXT NOT NULL DEFAULT (datetime('now'))
        );

        CREATE TABLE IF NOT EXISTS state_checkpoints (
            checkpoint_id TEXT PRIMARY KEY,
            checkpoint_name TEXT NOT NULL,
            state_root_hash TEXT NOT NULL,
            created_at TEXT NOT NULL DEFAULT (datetime('now'))
        );
        """)
    conn.close()

File: backend/src/database/connection.ts
import { DatabaseSync } from 'node:sqlite';
import { config } from '../config/index.js';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let dbInstance: any = null;

export function getDatabase(): any {
  if (!dbInstance) {
    const dbDir = path.dirname(config.databasePath);
    if (!fs.existsSync(dbDir)) {
      fs.mkdirSync(dbDir, { recursive: true });
    }

    dbInstance = new DatabaseSync(config.databasePath);
    try {
      dbInstance.exec('PRAGMA journal_mode = WAL;');
    } catch {
      // Ignore pragma error if memory db
    }
    runMigrations(dbInstance);
    seedInitialData(dbInstance);
  }
  return dbInstance;
}

function runMigrations(db: any) {
  const migrationsDir = path.resolve(__dirname, 'migrations');
  if (fs.existsSync(migrationsDir)) {
    const files = fs.readdirSync(migrationsDir).filter(f => f.endsWith('.sql')).sort();
    for (const file of files) {
      const sql = fs.readFileSync(path.join(migrationsDir, file), 'utf8');
      db.exec(sql);
    }
  }
}

function seedInitialData(db: any) {
  try {
    const row = db.prepare('SELECT count(*) as count FROM assets').get() as { count: number | bigint };
    if (row && Number(row.count) > 0) return;

    // 1. Sync State
    db.prepare(`
      INSERT OR REPLACE INTO sync_state (key, last_block, updated_at)
      VALUES ('corporate_actions', 194825, CURRENT_TIMESTAMP)
    `).run();

    // 2. Asset (Digital Bond Token)
    const assetAddress = '0x5FbDB2315678afecb367f032d93F642f64180aa3';
    db.prepare(`
      INSERT INTO assets (address, name, symbol, total_supply)
      VALUES (?, 'Digital Bond Token', 'DBT', '1000')
    `).run(assetAddress);

    // 3. Holders
    db.prepare(`
      INSERT INTO holders (address, asset_address, balance, is_whitelisted)
      VALUES (?, ?, '500', 1)
    `).run('0x70997970C51812dc3A010C7d01b50e0d17dc79C8', assetAddress); // Alice

    db.prepare(`
      INSERT INTO holders (address, asset_address, balance, is_whitelisted)
      VALUES (?, ?, '300', 1)
    `).run('0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC', assetAddress); // Bob

    db.prepare(`
      INSERT INTO holders (address, asset_address, balance, is_whitelisted)
      VALUES (?, ?, '200', 1)
    `).run('0x90F79bf6EB2c4f870365E785982E1f101E93b906', assetAddress); // Charlie

    // 4. Corporate Action CA-001
    const actionId = 'CA-001';
    const v1Id = '0x1a8f92b7c0d38119e48271829374019283740192837401928374019283740191';
    const v2Id = '0x2b9c81a4e1d49220f59382930485120394851203948512039485120394851202';

    db.prepare(`
      INSERT INTO corporate_actions (action_id, asset_address, action_type, active_version_id, status)
      VALUES (?, ?, 'COUPON_PAYMENT', ?, 'ACTIVE')
    `).run(actionId, assetAddress, v2Id);

    // Version 1 (SUPERSEDED)
    db.prepare(`
      INSERT INTO action_versions (
        version_id, action_id, version_number, rate_bps, amount_per_token,
        record_date, payable_date, document_hash, status, announced_by
      ) VALUES (?, ?, 1, 500, '0.05', 1789430400, 1789473600, 'QmXoypizjW3WknFiJnKLwHCnL72vedxjQkDDP1mXWo6uco', 'SUPERSEDED', '0x90F79bf6EB2c4f870365E785982E1f101E93b906')
    `).run(v1Id, actionId);

    // Version 2 (ACTIVE)
    db.prepare(`
      INSERT INTO action_versions (
        version_id, action_id, version_number, rate_bps, amount_per_token,
        record_date, payable_date, supersedes_version_id, document_hash, status, announced_by
      ) VALUES (?, ?, 2, 400, '0.04', 1789430400, 1789732800, ?, 'QmZtmD2qtW3wT1xYy72vedxjQkDD73hwo81kNmE9281kNm', 'ACTIVE', '0x90F79bf6EB2c4f870365E785982E1f101E93b906')
    `).run(v2Id, actionId, v1Id);

    // 5. Payment Events
    db.prepare(`
      INSERT INTO payment_events (action_id, version_id, holder_address, amount, asset_balance, transaction_hash, block_number, timestamp)
      VALUES (?, ?, '0x70997970C51812dc3A010C7d01b50e0d17dc79C8', '20.0', '500', '0x3c78a1f29d91827364bfa109823471029384710293847102938471029384710a', 194825, 1789732800)
    `).run(actionId, v2Id);

    db.prepare(`
      INSERT INTO payment_events (action_id, version_id, holder_address, amount, asset_balance, transaction_hash, block_number, timestamp)
      VALUES (?, ?, '0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC', '12.0', '300', '0x3c78a1f29d91827364bfa109823471029384710293847102938471029384710a', 194825, 1789732800)
    `).run(actionId, v2Id);

    db.prepare(`
      INSERT INTO payment_events (action_id, version_id, holder_address, amount, asset_balance, transaction_hash, block_number, timestamp)
      VALUES (?, ?, '0x90F79bf6EB2c4f870365E785982E1f101E93b906', '8.0', '200', '0x3c78a1f29d91827364bfa109823471029384710293847102938471029384710a', 194825, 1789732800)
    `).run(actionId, v2Id);

    // 6. Audit Logs
    const logStmt = db.prepare(`
      INSERT INTO audit_logs (event_type, action_id, version_id, actor, details, transaction_hash, block_number)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    `);

    logStmt.run('TokenMint', null, null, 'Genesis', JSON.stringify({ asset: 'Digital Bond Token', totalSupply: '1000 DBT' }), '0x2d81726354819283740192837401928374019283740192837401928374019281', 194801);
    logStmt.run('WhitelistUpdated', null, null, 'ComplianceOfficer', JSON.stringify({ accounts: ['Alice', 'Bob', 'Charlie'], status: 'APPROVED' }), '0x9c81729384710293847102938471029384710293847102938471029384710293', 194805);
    logStmt.run('ActionCreated', actionId, v1Id, 'Issuer', JSON.stringify({ actionId, version: 1, rateBps: 500, type: 'COUPON' }), '0x4f82719283740192837401928374019283740192837401928374019283740192', 194810);
    logStmt.run('Transfer', null, null, 'Bob', JSON.stringify({ from: 'Bob', to: 'Charlie', amount: '200 DBT' }), '0x7a81928374019283740192837401928374019283740192837401928374019283', 194819);
    logStmt.run('ActionAmended', actionId, v2Id, 'Issuer', JSON.stringify({ actionId, prevVersionId: v1Id, newVersionId: v2Id, newRateBps: 400 }), '0x1e82937401928374019283740192837401928374019283740192837401928371', 194821);
    logStmt.run('ActionPaymentExecuted', actionId, v2Id, 'PaymentExecutor', JSON.stringify({ actionId, version: 2, totalDisbursed: '40.00 USDC', holdersCount: 3 }), '0x3c78a1f29d91827364bfa109823471029384710293847102938471029384710a', 194825);
  } catch (err) {
    console.error('Seed data error:', err);
  }
}
